major rewrites to simplify new seasons

khari05 opened the pull request:
~~former title: update for 2021-2022~~

~~TODO:~~

~~- [x] fix building on android~~
~~- [x] finish the match breakdown~~
~~- [x] update localization~~
~~- [ ] maybe more descriptive member variable names for FreightFrenzyAllianceDetails?~~

~~I might get busy, here are my current changes.~~

New features:

Should be able to render match breakdowns using only i18n files and JSON schema

Dynamic dark theme w/ easy-dynamic-theme

TODO:
- [x] cleanup
- [x] add in the remaining seasons
- [x] add error catching for some of my precarious assumptions
- [x] Add back the unable to show data page

khari05 committed:
upgraded to android embedding v2

khari05 committed:
update dependencies

khari05 committed:
builds on ios now..
has some issue with android embedding v2 and firebase messaging

khari05 committed:
added parsing for the freight frenzy matches

khari05 committed:
turned off autocorrect on the login page

khari05 committed:
formatting

khari05 committed:
WIP need to finish the match breakdowns

khari05 committed:
removed leftover code from last year

khari05 committed:
update all match breakdowns for 2122

khari05 committed:
doesn't display remote breakdowns since the data is odd

khari05 commented:
Hey @ofekashery @theswerd, can you add HE and ES localization?

ofekashery committed:
Add FREIGHT FRENZY Hebrew translation

ofekashery committed:
Fix typo

nholo1332 commented:
I can add the Spanish translations. I'll try to get that done today

khari05 commented:
I'm currently trying to get the app to build on Android since that is the last step before we can release this version

ofekashery commented:
Thank you so much, @khari05!

nholo1332 commented:
Awesome, thanks! Once we're ready for this build to be released, I can push the iOS version to the AppStore.

nholo1332 committed:
Add Spanish translation

khari05 committed:
BUILDS ON ANDROID (regen android project/follow flutterfire upgrade guide)

khari05 committed:
yay this is finally implemented...
THEORETICALLY we no longer have to write new code every season!
TODO: ADD ERROR CHECKING; im sleepy

khari05 committed:
add TODOs, minor patches

khari05 commented:
I'm really happy with how straightforward it was to overhaul the most tedious part of updating the app

~~Only needed like 3 pages of infrastructure diagrams~~

khari05 committed:
made team pages open 2.35x faster...
by parallelizing the API requests so they complete asynchronously
fixed a small error w/ skystone matches containing some null value

khari05 commented:
In my rough testing, this previous commit makes the team info page open in half the time. Here are some numbers I collected from Dart DevTools:

| first request made | last request finishes | total time    | average        |
|--------------------|-----------------------|---------------|----------------|
| 17:29:21.878       | 17:29:30.791          | 9.03 seconds  |                |
| 17:35:57.557       | 17:36:06.513          | 8.956 seconds | 8.993 seconds  |
| 17:39:49.243       | 17:39:53.116          | 3.873 seconds |                |
| 17:46:59.292       | 17:47:03.080          | 3.788 seconds | 3.8305 seconds |

khari05 commented:
Hey @ofekashery can you take a look at this?

ofekashery commented on lib/models/ranking.dart:
```diff
@@ -39,30 +39,31 @@ class Ranking {
-        .toList()
-        .cast<Ranking>();
+    final list = List.from(json.decode(response));
+    list.removeWhere((e) => e == null); // WORKAROUND FOR SKYSTONE
```
Should be resolved now.

ofekashery commented on lib/ui/views/team/subpages/team-results.dart:
```diff
@@ -95,31 +103,36 @@ class TeamResults extends StatelessWidget {
 
     // Get all the team's rankings
-    List<Ranking> allRankings = await ApiV3().getTeamResults(teamKey, StaticData().seasonKey);
+    List<Ranking> allRankings = await ApiV3().getTeamResults(teamKey, StaticData.seasonKey);
+
+    final teamEvents = await ApiV3().getTeamEvents(teamKey, StaticData.seasonKey);
```
Can we use parallel loading for those?

ofekashery commented on lib/ui/views/team/subpages/team-results.dart:
```diff
@@ -95,31 +103,36 @@ class TeamResults extends StatelessWidget {
 
         // Get event detail
-        teamParticipant.event = await ApiV3().getEvent(eventParticipant.eventKey);
+        teamParticipant.event =
```
Please add to `requests`.

ofekashery commented on lib/ui/views/team/subpages/team-results.dart:
```diff
@@ -95,31 +103,36 @@ class TeamResults extends StatelessWidget {
+          List<Match> teamMatches = [];
           for (int i = 0; i < matches.length; i++) {
             Match match = matches[i];
             for (int i = 0; i < match.participants.length; i++) {
```
Can we rewrite it with `.where`?

khari05 replied:
definitely, I'm surprised I missed that!

ofekashery reviewed (changes requested):
Thank you @khari05! That's amazing.

khari05 committed:
parallelize more! its somehow twice as fast AGAIN!

khari05 committed:
fix awards + change match sort to put finals at the top

khari05 committed:
add in new season, remove redundant code
it took barely an hour to add in support for 2223

khari05 commented:
Using a bit of json schema and helper functions greatly reduces the code we need to write. I think it took me ≈3 hours to add in this past season.
<img width="483" alt="image" src="https://github.com/the-orange-alliance/mobile/assets/35894769/962a746d-2d50-4fd7-b4fd-fb00ec34dba3">

khari05 committed:
move custom breakdown rows to PowerPlayHelper

ofekashery commented:
Thank you so much @khari05 ❤️
